Have an e2e test against a node version that uses v8 very close to head

As CPU or heap profiler in v8 get any changes, we should continuously make sure those do not carry surprises for us (e.g. there is https://chromium-review.googlesource.com/c/v8/v8/+/1039825, which shouldn't break us but only green test gives confidence). To do that we need an automated e2e test run against a version of node that uses v8 very close to head. @ofrobots any recommendations on how to organize this? My naive hope would be that we could do something like "nvm install head" and the head gets v8 updated regularly (automatically?).

@nolanmar511 FYI

I would like to re-open this as I think using the Node.js canary build may be simple enough.

While verifying recent CPU profiler memory usage improvements, I used this command to install the [canary build](https://github.com/creationix/nvm/issues/1551):

```
NVM_NODEJS_ORG_MIRROR=https://nodejs.org/download/v8-canary nvm install node
```

With that I indeed had issues installing the profiler module:

```
$ npm install --save @google-cloud/profiler

> @google-cloud/profiler@0.2.0 install /tmp/testprof/node_modules/@google-cloud/profiler
> node-gyp rebuild

gyp WARN install got an error, rolling back install
gyp ERR! configure error 
gyp ERR! stack Error: 404 response downloading https://nodejs.org/download/v8-canary/v11.0.0-v8-canary201807140f69779e03/node-v11.0.0-v8-canary201807140f69779e03-headers.tar.gz
gyp ERR! stack     at Request.<anonymous> (/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/lib/node_modules/npm/node_modules/node-gyp/lib/install.js:210:14)
gyp ERR! stack     at Request.emit (events.js:187:15)
gyp ERR! stack     at Request.onRequestResponse (/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/lib/node_modules/npm/node_modules/request/request.js:1062:10)
gyp ERR! stack     at ClientRequest.emit (events.js:182:13)
gyp ERR! stack     at HTTPParser.parserOnIncomingClient (_http_client.js:554:21)
gyp ERR! stack     at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
gyp ERR! stack     at TLSSocket.socketOnData (_http_client.js:440:20)
gyp ERR! stack     at TLSSocket.emit (events.js:182:13)
gyp ERR! stack     at addChunk (_stream_readable.js:283:12)
gyp ERR! stack     at readableAddChunk (_stream_readable.js:264:11)
gyp ERR! System Linux 4.9.0-6-amd64
gyp ERR! command "/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/bin/node" "/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /tmp/testprof/node_modules/@google-cloud/profiler
gyp ERR! node -v v11.0.0-v8-canary201807140f69779e03
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
npm WARN enoent ENOENT: no such file or directory, open '/tmp/testprof/package.json'
npm WARN testprof No description
npm WARN testprof No repository field.
npm WARN testprof No README data
npm WARN testprof No license field.
npm WARN You are using a pre-release version of node and things may not work as expected

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @google-cloud/profiler@0.2.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @google-cloud/profiler@0.2.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /usr/local/google/home/aalexand/.npm/_logs/2018-07-20T17_55_23_203Z-debug.log
```

But using [this trick](https://github.com/nodejs/build/issues/626#issuecomment-406021579) makes things work:

```
$ npm install --save --nodedir=$(dirname $(dirname $(which node))) @google-cloud/profiler

> @google-cloud/profiler@0.2.0 install /tmp/testprof/node_modules/@google-cloud/profiler
> node-gyp rebuild

make: Entering directory '/tmp/testprof/node_modules/@google-cloud/profiler/build'
  CXX(target) Release/obj.target/sampling_heap_profiler/bindings/sampling-heap-profiler.o
  SOLINK_MODULE(target) Release/obj.target/sampling_heap_profiler.node
  COPY Release/sampling_heap_profiler.node
  CXX(target) Release/obj.target/time_profiler/bindings/time-profiler.o
../bindings/time-profiler.cc: In function ‘Nan::NAN_METHOD_RETURN_TYPE SetIdle(Nan::NAN_METHOD_ARGS_TYPE)’:
../bindings/time-profiler.cc:92:31: warning: ‘void v8::CpuProfiler::SetIdle(bool)’ is deprecated: Use Isolate::SetIdle(bool) instead. [-Wdeprecated-declarations]
   cpuProfiler->SetIdle(is_idle);
                               ^
In file included from /usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/include/node/v8.h:26:0,
                 from /usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/include/node/v8-profiler.h:10,
                 from ../bindings/time-profiler.cc:17:
/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/include/node/v8-profiler.h:360:22: note: declared here
                 void SetIdle(bool is_idle));
                      ^
/usr/local/google/home/aalexand/.nvm/versions/node/v11.0.0-v8-canary201807140f69779e03/include/node/v8config.h:324:3: note: in definition of macro ‘V8_DEPRECATED’
   declarator __attribute__((deprecated(message)))
   ^~~~~~~~~~
  SOLINK_MODULE(target) Release/obj.target/time_profiler.node
  COPY Release/time_profiler.node
make: Leaving directory '/tmp/testprof/node_modules/@google-cloud/profiler/build'

> protobufjs@6.8.8 postinstall /tmp/testprof/node_modules/protobufjs
> node scripts/postinstall

npm WARN saveError ENOENT: no such file or directory, open '/tmp/testprof/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/tmp/testprof/package.json'
npm WARN testprof No description
npm WARN testprof No repository field.
npm WARN testprof No README data
npm WARN testprof No license field.
npm WARN You are using a pre-release version of node and things may not work as expected

+ @google-cloud/profiler@0.2.0
added 117 packages from 155 contributors and audited 269 packages in 3.317s
found 0 vulnerabilities
```

And the headers in that location seems recent enough, at least they do include the recent line mode changes:

```
$ cat $(dirname $(dirname $(which node)))/include/node/v8-profiler.h | grep -C5 kCaller
  // (from the root to a leaf) will have line numbers that point to the start
  // line of the function, rather than the line of the callsite of the child.
  kLeafNodeLineNumbers,
  // In the resulting CpuProfile tree, nodes are separated based on the line
  // number of their callsite in their parent.
  kCallerLineNumbers,
};

/**
 * Interface for controlling CPU profiling. Instance of the
 * profiler can be created using v8::CpuProfiler::New method.
```

With that, adding testing against the canary build should be trivial?

The original crash has been solved, but still having problems with v8 canary.

Right now, when I try `NVM_NODEJS_ORG_MIRROR=https://nodejs.org/download/v8-canary nvm install node`, I get this error in the output:
```
../deps/v8/src/base/format-macros.h:24:20: fatal error: stddef.h: No such file or directory
 #include <stddef.h>
```

--- a/bindings/sampling-heap-profiler.cc
+++ b/bindings/sampling-heap-profiler.cc
@@ -61,8 +61,15 @@ NAN_METHOD(StartSamplingHeapProfiler) {
     if (!info[1]->IsNumber()) {
       return Nan::ThrowTypeError("First argument type must be Integer.");
     }
+
+    #if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
+    uint64_t sample_interval = info[0].As<Integer>()->Value();
+    int stack_depth = info[1].As<Integer>()->Value();
+    #else
     uint64_t sample_interval = info[0].As<Integer>()->Uint32Value();
     int stack_depth = info[1].As<Integer>()->IntegerValue();
+    #endif
+
     info.GetIsolate()->GetHeapProfiler()->
       StartSamplingHeapProfiler(sample_interval, stack_depth);
   } else {

--- a/bindings/sampling-heap-profiler.cc
+++ b/bindings/sampling-heap-profiler.cc
@@ -62,13 +62,13 @@ NAN_METHOD(StartSamplingHeapProfiler) {
       return Nan::ThrowTypeError("First argument type must be Integer.");
     }
 
-    #if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
+#if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
     uint64_t sample_interval = info[0].As<Integer>()->Value();
     int stack_depth = info[1].As<Integer>()->Value();
-    #else
+#else
     uint64_t sample_interval = info[0].As<Integer>()->Uint32Value();
     int stack_depth = info[1].As<Integer>()->IntegerValue();
-    #endif
+#endif
 
     info.GetIsolate()->GetHeapProfiler()->
       StartSamplingHeapProfiler(sample_interval, stack_depth);

--- a/bindings/time-profiler.cc
+++ b/bindings/time-profiler.cc
@@ -70,7 +70,7 @@ NAN_METHOD(StartProfiling) {
 
   // Sample counts and timestamps are not used, so we do not need to record
   // samples.
-  cpuProfiler->StartProfiling(name, false);
+  cpuProfiler->StartProfiling(name);
 }
 
 NAN_METHOD(StopProfiling) {
@@ -83,7 +83,11 @@ NAN_METHOD(StopProfiling) {
 }
 
 NAN_METHOD(SetSamplingInterval) {
+  #if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
+  int us = info[0].As<Integer>()->Value();
+  #else
   int us = info[0].As<Integer>()->IntegerValue();
+  #endif
   cpuProfiler->SetSamplingInterval(us);
 }
 

--- a/bindings/time-profiler.cc
+++ b/bindings/time-profiler.cc
@@ -70,7 +70,7 @@ NAN_METHOD(StartProfiling) {
 
   // Sample counts and timestamps are not used, so we do not need to record
   // samples.
-  cpuProfiler->StartProfiling(name);
+  cpuProfiler->StartProfiling(name, false);
 }
 
 NAN_METHOD(StopProfiling) {
@@ -83,11 +83,11 @@ NAN_METHOD(StopProfiling) {
 }
 
 NAN_METHOD(SetSamplingInterval) {
-  #if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
+#if NODE_MODULE_VERSION > NODE_10_0_MODULE_VERSION
   int us = info[0].As<Integer>()->Value();
-  #else
+#else
   int us = info[0].As<Integer>()->IntegerValue();
-  #endif
+#endif
   cpuProfiler->SetSamplingInterval(us);
 }
 

--- a/package.json
+++ b/package.json
@@ -34,7 +34,7 @@
     "delay": "^4.0.1",
     "extend": "^3.0.1",
     "gcp-metadata": "^0.8.0",
-    "nan": "^2.8.0",
+    "nan": "^2.11.1",
     "parse-duration": "^0.1.1",
     "pify": "^4.0.0",
     "pretty-ms": "^4.0.0",

--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "name": "@google-cloud/profiler",
   "version": "0.2.2",
   "description": "Adds support for Stackdriver Profiler to node.js applications",
-  "repository": "GoogleCloudPlatform/cloud-profiler-nodejs",
+  "repository": "googleapis/cloud-profiler-nodejs",
   "main": "out/src/index.js",
   "types": "out/src/index.d.ts",
   "scripts": {

--- a/testing/integration_test.go
+++ b/testing/integration_test.go
@@ -74,9 +74,10 @@ export NVM_DIR="$HOME/.nvm" >/dev/null
 
 # nvm install writes to stderr and stdout on successful install, so both are
 # redirected.
-retry nvm install {{.NodeVersion}} &>/dev/null
+{{if .NVMMirror}}NVM_NODEJS_ORG_MIRROR={{.NVMMirror}}{{end}} retry nvm install {{.NodeVersion}} &>/dev/null
 npm -v
 node -v
+NODEDIR=$(dirname $(dirname $(which node)))
 
 # Install agent
 retry git clone {{.Repo}}
@@ -84,9 +85,9 @@ cd cloud-profiler-nodejs
 retry git fetch origin {{if .PR}}pull/{{.PR}}/head{{else}}{{.Branch}}{{end}}:pull_branch
 git checkout pull_branch
 git reset --hard {{.Commit}}
-retry npm install >/dev/null
-npm run compile
-npm pack >/dev/null
+retry npm install --nodedir="$NODEDIR" >/dev/null
+npm run compile 
+npm pack --nodedir="$NODEDIR" >/dev/null
 VERSION=$(node -e "console.log(require('./package.json').version);")
 PROFILER="$HOME/cloud-profiler-nodejs/google-cloud-profiler-$VERSION.tgz"
 
@@ -95,7 +96,7 @@ mkdir -p "$TESTDIR"
 cp -r "testing/busybench" "$TESTDIR"
 cd "$TESTDIR/busybench"
 
-retry npm install "$PROFILER" >/dev/null
+retry npm install --nodedir="$NODEDIR" "$PROFILER" >/dev/null
 retry npm install
 npm run compile
 
@@ -118,6 +119,7 @@ type nodeGCETestCase struct {
 	proftest.InstanceConfig
 	name         string
 	nodeVersion  string
+	nvmMirror    string
 	wantProfiles []profileSummary
 }
 
@@ -127,6 +129,7 @@ func (tc *nodeGCETestCase) initializeStartUpScript(template *template.Template) 
 		struct {
 			Service     string
 			NodeVersion string
+			NVMMirror   string
 			Repo        string
 			PR          int
 			Branch      string
@@ -134,6 +137,7 @@ func (tc *nodeGCETestCase) initializeStartUpScript(template *template.Template) 
 		}{
 			Service:     tc.name,
 			NodeVersion: tc.nodeVersion,
+			NVMMirror:   tc.nvmMirror,
 			Repo:        *repo,
 			PR:          *pr,
 			Branch:      *branch,
@@ -229,6 +233,18 @@ func TestAgentIntegration(t *testing.T) {
 			name:         fmt.Sprintf("profiler-test-node10-%s-gce", runID),
 			wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
 			nodeVersion:  "10",
+		},
+		{
+			InstanceConfig: proftest.InstanceConfig{
+				ProjectID:   projectID,
+				Zone:        zone,
+				Name:        fmt.Sprintf("profiler-test-v8-canary-%s", runID),
+				MachineType: "n1-standard-1",
+			},
+			name:         fmt.Sprintf("profiler-test-v8-canary-%s-gce", runID),
+			wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
+			nodeVersion:  "node", // install latest version of node
+			nvmMirror:    "https://nodejs.org/download/v8-canary",
 		},
 	}
 

--- a/testing/integration_test.go
+++ b/testing/integration_test.go
@@ -34,12 +34,13 @@ import (
 )
 
 var (
-	repo        = flag.String("repo", "https://github.com/GoogleCloudPlatform/cloud-profiler-nodejs.git", "git repo to test")
-	branch      = flag.String("branch", "", "git branch to test")
-	commit      = flag.String("commit", "", "git commit to test")
-	pr          = flag.Int("pr", 0, "git pull request to test")
-	runID       = strings.Replace(time.Now().Format("2006-01-02-15-04-05.000000-0700"), ".", "-", -1)
-	runV8Canary = flag.Bool("v8_canary", false, "if tests should be run with the v8 canary build")
+	repo               = flag.String("repo", "https://github.com/googleapis/cloud-profiler-nodejs.git", "git repo to test")
+	branch             = flag.String("branch", "", "git branch to test")
+	commit             = flag.String("commit", "", "git commit to test")
+	pr                 = flag.Int("pr", 0, "git pull request to test")
+	enableV8CanaryTest = flag.Bool("enable_v8_canary_test", false, "if tests run with the v8-canary build of node should be enabled")
+
+	runID = strings.Replace(time.Now().Format("2006-01-02-15-04-05.000000-0700"), ".", "-", -1)
 )
 
 const cloudScope = "https://www.googleapis.com/auth/cloud-platform"
@@ -151,7 +152,45 @@ func (tc *nodeGCETestCase) initializeStartUpScript(template *template.Template) 
 	return nil
 }
 
-func gceTestcases(projectID, zone string, runV8Canary bool) []nodeGCETestCase {
+func TestAgentIntegration(t *testing.T) {
+	projectID := os.Getenv("GCLOUD_TESTS_NODEJS_PROJECT_ID")
+	if projectID == "" {
+		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_PROJECT_ID) got empty string")
+	}
+
+	zone := os.Getenv("GCLOUD_TESTS_NODEJS_ZONE")
+	if zone == "" {
+		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_ZONE) got empty string")
+	}
+
+	if *commit == "" {
+		t.Fatal("commit flag is not set")
+	}
+
+	ctx := context.Background()
+
+	client, err := google.DefaultClient(ctx, cloudScope)
+	if err != nil {
+		t.Fatalf("failed to get default client: %v", err)
+	}
+
+	computeService, err := compute.New(client)
+	if err != nil {
+		t.Fatalf("failed to initialize compute Service: %v", err)
+	}
+
+	template, err := template.New("startupScript").Parse(startupTemplate)
+	if err != nil {
+		t.Fatalf("failed to parse startup script template: %v", err)
+	}
+
+	gceTr := proftest.GCETestRunner{
+		TestRunner: proftest.TestRunner{
+			Client: client,
+		},
+		ComputeService: computeService,
+	}
+
 	testcases := []nodeGCETestCase{
 		{
 			InstanceConfig: proftest.InstanceConfig{
@@ -198,64 +237,22 @@ func gceTestcases(projectID, zone string, runV8Canary bool) []nodeGCETestCase {
 			nodeVersion:  "10",
 		},
 	}
-	if !runV8Canary {
-		return testcases
-	}
-	v8CanaryTC := nodeGCETestCase{
-		InstanceConfig: proftest.InstanceConfig{
-			ProjectID:   projectID,
-			Zone:        zone,
-			Name:        fmt.Sprintf("profiler-test-v8-canary-%s", runID),
-			MachineType: "n1-standard-1",
-		},
-		name:         fmt.Sprintf("profiler-test-v8-canary-%s-gce", runID),
-		wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
-		nodeVersion:  "node", // install latest version of node
-		nvmMirror:    "https://nodejs.org/download/v8-canary",
-	}
-	return append(testcases, v8CanaryTC)
-}
-
-func TestAgentIntegration(t *testing.T) {
-	projectID := os.Getenv("GCLOUD_TESTS_NODEJS_PROJECT_ID")
-	if projectID == "" {
-		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_PROJECT_ID) got empty string")
+	if *enableV8CanaryTest {
+		testcases := append(testcases,
+			nodeGCETestCase{
+				InstanceConfig: proftest.InstanceConfig{
+					ProjectID:   projectID,
+					Zone:        zone,
+					Name:        fmt.Sprintf("profiler-test-v8-canary-%s", runID),
+					MachineType: "n1-standard-1",
+				},
+				name:         fmt.Sprintf("profiler-test-v8-canary-%s-gce", runID),
+				wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
+				nodeVersion:  "node", // install latest version of node
+				nvmMirror:    "https://nodejs.org/download/v8-canary",
+			})
 	}
 
-	zone := os.Getenv("GCLOUD_TESTS_NODEJS_ZONE")
-	if zone == "" {
-		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_ZONE) got empty string")
-	}
-
-	if *commit == "" {
-		t.Fatal("commit flag is not set")
-	}
-
-	ctx := context.Background()
-
-	client, err := google.DefaultClient(ctx, cloudScope)
-	if err != nil {
-		t.Fatalf("failed to get default client: %v", err)
-	}
-
-	computeService, err := compute.New(client)
-	if err != nil {
-		t.Fatalf("failed to initialize compute Service: %v", err)
-	}
-
-	template, err := template.New("startupScript").Parse(startupTemplate)
-	if err != nil {
-		t.Fatalf("failed to parse startup script template: %v", err)
-	}
-
-	gceTr := proftest.GCETestRunner{
-		TestRunner: proftest.TestRunner{
-			Client: client,
-		},
-		ComputeService: computeService,
-	}
-
-	testcases := gceTestcases(projectID, zone, *runV8Canary)
 	// Allow test cases to run in parallel.
 	runtime.GOMAXPROCS(len(testcases))
 

--- a/testing/integration_test.go
+++ b/testing/integration_test.go
@@ -34,11 +34,12 @@ import (
 )
 
 var (
-	repo   = flag.String("repo", "https://github.com/GoogleCloudPlatform/cloud-profiler-nodejs.git", "git repo to test")
-	branch = flag.String("branch", "", "git branch to test")
-	commit = flag.String("commit", "", "git commit to test")
-	pr     = flag.Int("pr", 0, "git pull request to test")
-	runID  = strings.Replace(time.Now().Format("2006-01-02-15-04-05.000000-0700"), ".", "-", -1)
+	repo        = flag.String("repo", "https://github.com/GoogleCloudPlatform/cloud-profiler-nodejs.git", "git repo to test")
+	branch      = flag.String("branch", "", "git branch to test")
+	commit      = flag.String("commit", "", "git commit to test")
+	pr          = flag.Int("pr", 0, "git pull request to test")
+	runID       = strings.Replace(time.Now().Format("2006-01-02-15-04-05.000000-0700"), ".", "-", -1)
+	runV8Canary = flag.Bool("v8_canary", false, "if tests should be run with the v8 canary build")
 )
 
 const cloudScope = "https://www.googleapis.com/auth/cloud-platform"
@@ -150,45 +151,7 @@ func (tc *nodeGCETestCase) initializeStartUpScript(template *template.Template) 
 	return nil
 }
 
-func TestAgentIntegration(t *testing.T) {
-	projectID := os.Getenv("GCLOUD_TESTS_NODEJS_PROJECT_ID")
-	if projectID == "" {
-		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_PROJECT_ID) got empty string")
-	}
-
-	zone := os.Getenv("GCLOUD_TESTS_NODEJS_ZONE")
-	if zone == "" {
-		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_ZONE) got empty string")
-	}
-
-	if *commit == "" {
-		t.Fatal("commit flag is not set")
-	}
-
-	ctx := context.Background()
-
-	client, err := google.DefaultClient(ctx, cloudScope)
-	if err != nil {
-		t.Fatalf("failed to get default client: %v", err)
-	}
-
-	computeService, err := compute.New(client)
-	if err != nil {
-		t.Fatalf("failed to initialize compute Service: %v", err)
-	}
-
-	template, err := template.New("startupScript").Parse(startupTemplate)
-	if err != nil {
-		t.Fatalf("failed to parse startup script template: %v", err)
-	}
-
-	gceTr := proftest.GCETestRunner{
-		TestRunner: proftest.TestRunner{
-			Client: client,
-		},
-		ComputeService: computeService,
-	}
-
+func gceTestcases(projectID, zone string, runV8Canary bool) []nodeGCETestCase {
 	testcases := []nodeGCETestCase{
 		{
 			InstanceConfig: proftest.InstanceConfig{
@@ -234,20 +197,65 @@ func TestAgentIntegration(t *testing.T) {
 			wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
 			nodeVersion:  "10",
 		},
-		{
-			InstanceConfig: proftest.InstanceConfig{
-				ProjectID:   projectID,
-				Zone:        zone,
-				Name:        fmt.Sprintf("profiler-test-v8-canary-%s", runID),
-				MachineType: "n1-standard-1",
-			},
-			name:         fmt.Sprintf("profiler-test-v8-canary-%s-gce", runID),
-			wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
-			nodeVersion:  "node", // install latest version of node
-			nvmMirror:    "https://nodejs.org/download/v8-canary",
+	}
+	if !runV8Canary {
+		return testcases
+	}
+	v8CanaryTC := nodeGCETestCase{
+		InstanceConfig: proftest.InstanceConfig{
+			ProjectID:   projectID,
+			Zone:        zone,
+			Name:        fmt.Sprintf("profiler-test-v8-canary-%s", runID),
+			MachineType: "n1-standard-1",
 		},
+		name:         fmt.Sprintf("profiler-test-v8-canary-%s-gce", runID),
+		wantProfiles: []profileSummary{{"WALL", "busyLoop"}, {"HEAP", "benchmark"}},
+		nodeVersion:  "node", // install latest version of node
+		nvmMirror:    "https://nodejs.org/download/v8-canary",
+	}
+	return append(testcases, v8CanaryTC)
+}
+
+func TestAgentIntegration(t *testing.T) {
+	projectID := os.Getenv("GCLOUD_TESTS_NODEJS_PROJECT_ID")
+	if projectID == "" {
+		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_PROJECT_ID) got empty string")
 	}
 
+	zone := os.Getenv("GCLOUD_TESTS_NODEJS_ZONE")
+	if zone == "" {
+		t.Fatalf("Getenv(GCLOUD_TESTS_NODEJS_ZONE) got empty string")
+	}
+
+	if *commit == "" {
+		t.Fatal("commit flag is not set")
+	}
+
+	ctx := context.Background()
+
+	client, err := google.DefaultClient(ctx, cloudScope)
+	if err != nil {
+		t.Fatalf("failed to get default client: %v", err)
+	}
+
+	computeService, err := compute.New(client)
+	if err != nil {
+		t.Fatalf("failed to initialize compute Service: %v", err)
+	}
+
+	template, err := template.New("startupScript").Parse(startupTemplate)
+	if err != nil {
+		t.Fatalf("failed to parse startup script template: %v", err)
+	}
+
+	gceTr := proftest.GCETestRunner{
+		TestRunner: proftest.TestRunner{
+			Client: client,
+		},
+		ComputeService: computeService,
+	}
+
+	testcases := gceTestcases(projectID, zone, *runV8Canary)
 	// Allow test cases to run in parallel.
 	runtime.GOMAXPROCS(len(testcases))
 

--- a/testing/integration_test.sh
+++ b/testing/integration_test.sh
@@ -33,7 +33,7 @@ cp -R "testing" "$GOPATH/src/proftest"
 cd "$GOPATH/src/proftest"
 retry go get -t -tags=integration .
 if [ "$KOKORO_GITHUB_PULL_REQUEST_NUMBER" == "" ]; then
-  go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -branch="$BRANCH" -repo="$REPO"
+  go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -branch="$BRANCH" -repo="$REPO" -v8_canary
 else 
   go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -pr="$KOKORO_GITHUB_PULL_REQUEST_NUMBER"
 fi

--- a/testing/integration_test.sh
+++ b/testing/integration_test.sh
@@ -33,7 +33,7 @@ cp -R "testing" "$GOPATH/src/proftest"
 cd "$GOPATH/src/proftest"
 retry go get -t -tags=integration .
 if [ "$KOKORO_GITHUB_PULL_REQUEST_NUMBER" == "" ]; then
-  go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -branch="$BRANCH" -repo="$REPO" -v8_canary
+  go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -branch="$BRANCH" -repo="$REPO" -enable_v8_canary_test
 else 
   go test -timeout=30m -tags=integration -run TestAgentIntegration -commit="$COMMIT" -pr="$KOKORO_GITHUB_PULL_REQUEST_NUMBER"
 fi

